File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        zeros = 0
        for num in nums:
            if num==0:
                zeros += 1
        ans = [0] * (n)
        if zeros > 1:
            return ans
        ans[0] = 1
        for i in range(1,n):
            ans[i] = ans[i-1] * nums[i-1]
        tmp = nums[n-1]
        for i in range(n-2, -1, -1):
            ans[i] *= tmp
            tmp *= nums[i]
        return ans

File: Data Structures & Algorithms/top-k-elements-in-list/submission-3.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        int n = nums.length;
        List<List<Integer>> freqArr = new ArrayList<>();
        Map<Integer, Integer> hm = new HashMap<>();
        int[] ans = new int[k];
        for(int i=0; i<n; i++) {
            freqArr.add(new ArrayList<>());
            hm.put(nums[i], hm.getOrDefault(nums[i],0)+1);
        } // freq 1 2 3 - - -
        for(int key : hm.keySet()) {
            freqArr.get(hm.get(key)-1).add(key);
        }
        int idx = 0;
        for(int i=n-1; i>=0; i--) {
            List<Integer> tmp = freqArr.get(i);
            for(int x : tmp) {
                ans[idx++] = x;
                if(idx==k) return ans;
            }
        }
        return ans;
    }
}
